controllers/k8ssandracluster_controller.go

```diff
@@ -215,23 +220,64 @@ func (r *K8ssandraClusterReconciler) reconcile(ctx context.Context, kc *api.K8ss
 				return ctrl.Result{}, err
 			}
 		}
+	}
 
-		if kc.HasStargates() && i == 0 {
-			// only needs to be done once per cluster, so we do it when reconciling the first dc only
-			if err := r.ensureStargateAuthKeyspaceExists(ctx, kc, actualDc, remoteClient, logger); err != nil {
-				return ctrl.Result{RequeueAfter: longDelay}, err
-			}
+	for _, dcTemplate := range kc.Spec.Cassandra.Datacenters {
+		dcConfig := cassandra.Coalesce(kc.Spec.Cassandra.DeepCopy(), dcTemplate.DeepCopy())
+		dcConfig.AdditionalSeeds = seeds
+		desiredDc, err := cassandra.NewDatacenter(kcKey, dcConfig)
+		dcKey := types.NamespacedName{Namespace: desiredDc.Namespace, Name: desiredDc.Name}
+		logger := kcLogger.WithValues("CassandraDatacenter", dcKey)
+
+		if err != nil {
+			logger.Error(err, "Failed to create new CassandraDatacenter")
+			return ctrl.Result{}, err
 		}
-		result, err := r.reconcileStargate(ctx, kc, dcTemplate, actualDc, logger, remoteClient)
-		if !result.IsZero() || err != nil {
-			return result, err
+
+		desiredDcHash := utils.DeepHashString(desiredDc)
+		desiredDc.Annotations[api.ResourceHashAnnotation] = desiredDcHash
+
+		actualDc := &cassdcapi.CassandraDatacenter{}
```

Comment:
I think it's a bit disappointing to re-fetch the actual DC here. It also might be slightly incorrect from the reconciliation logic perspective, because the DC specs could have been altered between the first fetch and the second one. I think it would be safer to store the actual DC in a slice:

```go
actualDcs := make([]*cassdcapi.CassandraDatacenter, 0, len(kc.Spec.Cassandra.Datacenters))
for i, dcTemplate := range kc.Spec.Cassandra.Datacenters {
    ...
    actualDcs[i] = actualDc
}
```

Then the reconcile loop for Stargate could be simplified:

```go
for i, dcTemplate := range kc.Spec.Cassandra.Datacenters {
	actualDc := actualDcs[i]
	dcKey := types.NamespacedName{Namespace: actualDc.Namespace, Name: actualDc.Name}
	logger := kcLogger.WithValues("CassandraDatacenter", dcKey)
	if remoteClient, err := r.buildRemoteClient(logger, dcTemplate, ctx, dcKey, actualDc, kc); err != nil {
		return ctrl.Result{}, err
	} else if result, err := r.reconcileStargate(ctx, kc, dcTemplate, actualDc, logger, remoteClient); !result.IsZero() || err != nil {
		return result, err
	}
}
```

Reply:
I was thinking about this but wasn't sure about the lifecycle of these objects.
I'll give your solution a try to see how it works.